Forward @:nativeTypeConstraints in closures to class filter

kLabz opened the pull request:
I have some issues with native cs libs using type parameter constraints: while I can generally use them with `@:nativeTypeConstraints` (see #3526), the closures to class filter generates classes without these type constraints, resulting in a successful haxe compilation but a compilation error in C#.

This PR does two things:
- forward `@:nativeTypeConstraints` meta to these classes
- add type parameter constraints to the needed type parameters

Do not hesitate to tell me if my code doesn't make sense or if there are better ways to do it; I'm still not familiar with ocaml / haxe sources.

Related question: should `@:nativeTypeConstraints` become a proper meta? (I could add it to this PR)

kLabz committed:
Forward @:nativeTypeConstraints in closures to class filter

nadako commented:
I am not really familiar (anymore) with this code, and I can't say I fully understand what's going on here, but forwarding constraints makes sense to me and since the tests still pass, it should be fine ;-)

Could you maybe add a test for this somehow?

nadako commented on src/codegen/gencommon/closuresToClass.ml:
```diff
@@ -413,6 +414,12 @@ let configure gen ft =
 		if in_unsafe then cls.cl_meta <- (Meta.Unsafe,[],null_pos) :: cls.cl_meta;
 
+		(* Forward native type constraints meta *)
+		if Meta.has(Meta.Custom(":nativeTypeConstraints")) (get gen.gcurrent_class).cl_meta then begin
```
I think it's better to also do the `is_hxgen` check here, or even better move `not is_hxgen && Meta.has` into a function and use it in both places. Also here you might want to check the platform, because this filter is also run on java (tho adding metadata should have no effect, but still).

kLabz replied:
Ok thanks, I'll look into this (not sure where to put such a function though).

nadako commented on src/codegen/gencommon/closuresToClass.ml:
```diff
@@ -278,8 +278,9 @@ let traverse gen ?tparam_anon_decl ?tparam_anon_acc (handle_anon_func:texpr->tfu
-			if List.memq cl acc then acc else cl :: acc
+		| TInst(( { cl_kind = KTypeParameter constraints } as cl), []) ->
+			let params = List.fold_left get_type_params acc constraints in
+			List.filter (fun t -> not (List.memq t acc)) (cl :: params) @ acc;
```
I don't really understand what's going on here and what's the purpose of this function. I see it basically collects all type params, but what this change actually does? Perhaps if you have a good understanding maybe you could comment it a bit for future generations? :)))

kLabz replied:
It may well be nonsense and only work by chance in my tests, I'm really not familiar with ocaml yet.

It collects all type params and runs again `get_type_params` on their constraints to be sure to have all necessary types. In the second line, it adds only types that are not already in `acc`, as before (except that now it runs that on both `cl` and `params` instead of just `cl`).

nadako replied:
Yeah I understood that far :) I'm just wondering how is this used after. But oh well, I guess I'll have to spend an evening with a camomile tea to get an idea of what's going on with type params in gencommon in general :)

kLabz commented:
> Could you maybe add a test for this somehow?

Sure, I'll look into this :)

kLabz committed:
Remove :nativeTypeConstraints custom meta

kLabz commented:
I actually dropped the `@:nativeTypeConstraints` custom meta and used `@:nativeGen` directly instead. Issues impacted: #3526 #4002 and potentially [hxcs#37](https://github.com/HaxeFoundation/hxcs/issues/37).

I'll add some tests, trying to reflect what was happening in my application.

kLabz committed:
Add test for #7875

kLabz committed:
Add another test for #4002

Note: this test was already passing with `@:nativeTypeConstraints`
on class Test, but this custom meta is being removed.

kLabz commented:
Tests have been added and CI passed.

The code related to metadata has been rewritten, you might want to review again.

kLabz commented:
@nadako ?

kLabz commented:
@Simn could this be considered for 4.0?

We would like to switch to official haxe 4.0 when it is available and possibly stick with it for a while for production apps, but our backend cannot work without this fix (or something similar).

I can work on this again as much as needed.

_Edit: just rebased haxe 4 rc2 to get an updated build from appveyor (hopefully) before monday._
